cpp/src/parquet/reader_test.cc

```diff
@@ -127,6 +127,89 @@ void CheckRowGroupMetadata(const RowGroupMetaData* rg_metadata,
   }
 }
 
+class TestBooleanRLE : public ::testing::Test {
+ public:
+  void SetUp() {
+    reader_ = ParquetFileReader::OpenFile(data_file("rle_boolean_encoding.parquet"));
+  }
+
+  void TearDown() {}
+
+ protected:
+  std::unique_ptr<ParquetFileReader> reader_;
+};
+
+TEST_F(TestBooleanRLE, TestBooleanScanner) {
+  auto group = reader_->RowGroup(0);
+
+  // column 0, id
+  auto scanner = std::make_shared<BoolScanner>(group->Column(0));
+
+  bool val = false;
+  bool is_null = false;
+  for (int i = 0; i < 8; i++) {
+    ASSERT_TRUE(scanner->HasNext());
+    ASSERT_TRUE(scanner->NextValue(&val, &is_null));
+
+    // For this file, 3rd index value is null
+    if (i == 2) {
+      ASSERT_TRUE(is_null);
+    } else {
+      ASSERT_FALSE(is_null);
+    }
+  }
+
+  ASSERT_FALSE(scanner->HasNext());
+  ASSERT_FALSE(scanner->NextValue(&val, &is_null));
+}
+
+TEST_F(TestBooleanRLE, TestBatchRead) {
+  auto group = reader_->RowGroup(0);
+
+  // column 0, id
+  auto col = std::dynamic_pointer_cast<BoolReader>(group->Column(0));
+
+  // This file only has 8 rows
+  ASSERT_EQ(8, reader_->metadata()->num_rows());
+  // This file only has 1 row group
+  ASSERT_EQ(1, reader_->metadata()->num_row_groups());
+  // Size of the metadata is 106 bytes
+  ASSERT_EQ(106, reader_->metadata()->size());
+  // This row group must have 8 rows
+  ASSERT_EQ(8, group->metadata()->num_rows());
+
+  // Check if the column is encoded with RLE
+  auto col_chunk = group->metadata()->ColumnChunk(0);
+  ASSERT_TRUE(std::find(col_chunk->encodings().begin(), col_chunk->encodings().end(),
+                        Encoding::RLE) != col_chunk->encodings().end());
+
+  // Assert column has values to be read
+  ASSERT_TRUE(col->HasNext());
+  int64_t curr_batch_read = 0;
+
+  const int16_t batch_size = 8;
+  int16_t def_levels[batch_size];
+  int16_t rep_levels[batch_size];
+  bool values[batch_size - 1];
+
+  auto levels_read =
+      col->ReadBatch(batch_size, def_levels, rep_levels, values, &curr_batch_read);
+  ASSERT_EQ(batch_size, levels_read);
+
+  // Since one value is a null value, expect batches read to be one less than indicated
+  // batch_size
+  ASSERT_EQ(batch_size - 1, curr_batch_read);
+
+  // 3rd index is null value
+  ASSERT_THAT(def_levels, testing::ElementsAre(1, 1, 0, 1, 1, 1, 1, 1));
+
+  // Validate inserted data is as expected
+  ASSERT_THAT(values, testing::ElementsAre(1, 0, 1, 1, 0, 0, 1));
```

Comment:
Does the input have both `<bit-packed-run>` and `<rle-run>` cases in https://parquet.apache.org/docs/file-format/data-pages/encodings/#a-namerlearun-length-encoding--bit-packing-hybrid-rle--3 ?

Reply:
Yes, the input has both `<bit-packed-run>` and `<rle-run>`.

Reply:
I used `gdb` and watched `RleDecoder::NextCounts()`. All calls use the `is_literal` clause:

https://github.com/apache/arrow/blob/8daa7a4ed5629c0020dadf7325a6b523bdfc62e9/cpp/src/arrow/util/rle_encoding.h#L653-L656

It seems that the test input have only `<bit-packed-run>` case.

Are all boolean values (1/0) use `<bit-packed-run>` because they can be represented with 1 bit?

Reply:
Hmmm, my bad. I thought the test case had both. 
No, Boolean values use both` bit-packed-run` and `rle-run`. 

I validated against another file : [rle_boolean_encoding.parquet](https://github.com/sfc-gh-nthimmegowda/parquet-testing/blob/master/data/rle_boolean_encoding.parquet). I could find both `is_literal` as true and false. 

I beleive since the test file contains only some data (8 rows) it could not go into `<rle-run>` , however large amount of rows will move it. 
My tentative plan is to replace the current test file with a new one with more rows. I would have done it in this PR, but the test files come from a submodule which require another PR.

Reply:
OK. Could you update the test file in https://github.com/apache/parquet-testing and then update the `cpp/submodules/parquet-testing/` in this pull request?

Reply:
Ack. Updated test file file in [parquet-testing](https://github.com/apache/parquet-testing) and updating in this pull request.